### PR TITLE
PB103 Migration

### PR DESCRIPTION
**Pull request by AradCarmi**
## Description
Added the PB103 to the new validation format. Validate whether there is an unconnected task.

### CODE REVIEW

File: demisto_sdk/commands/validate/sdk_validation_config.toml [hunk 1]
```diff
@@ -29,7 +29,7 @@ select = [
     "IN100", "IN101", "IN102", "IN104", "IN106", "IN107", "IN108", "IN109", "IN110", "IN112", "IN113", "IN114", "IN115", "IN117", "IN118", "IN121", "IN122", "IN123", "IN124", "IN125", "IN126", "IN127", "IN130",
     "IN131", "IN134", "IN135", "IN139", "IN141", "IN142", "IN144", "IN145", "IN146", "IN149", "IN150", "IN151", "IN152", "IN153", "IN154", "IN156", "IN158", "IN159", "IN160",
     "IN161", "IN162",
-    "PB100", "PB101","PB104","PB118", "PB123",
+    "PB100", "PB101","PB103","PB104","PB118", "PB123",
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/sdk_validation_config.toml:32]: Let's also add it under **path_based_validations**.

File: demisto_sdk/commands/validate/validators/PB_validators/PB103_is_playbook_has_unconnected_tasks.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            if task_id != start_task_id:
+                tasks_bucket.add(task_id)
+            next_tasks = task.get("nexttasks", {})
+            for next_task_ids in next_tasks.values():
+                if next_task_ids:
+                    next_tasks_bucket.update(next_task_ids)
+        orphan_tasks = tasks_bucket.difference(next_tasks_bucket)
+        if orphan_tasks:
+            return False
+        return tasks_bucket.issubset(next_tasks_bucket)
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/validators/PB_validators/PB103_is_playbook_has_unconnected_tasks.py:49]:
```suggestion
        orphan_tasks = tasks_bucket.difference(next_tasks_bucket)
        return tasks_bucket.issubset(next_tasks_bucket)
```
I think that's sufficient.

File: demisto_sdk/commands/validate/validators/PB_validators/PB103_is_playbook_has_unconnected_tasks.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    ValidationResult,
+)
+
+ContentTypes = Playbook
+
+
+class IsPlaybookHasUnconnectedTasks(BaseValidator[ContentTypes]):
+    error_code = "PB103"
+    description = "Validate whether there is an unconnected task."
+    rationale = "Validate whether there is an unconnected task to the root task."
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/validators/PB_validators/PB103_is_playbook_has_unconnected_tasks.py:17]:
```suggestion
    rationale = "Make sure there're no unconnected tasks to ensure the playbook will work as expected."
```
rationale should be something like this.

File: demisto_sdk/commands/validate/validators/PB_validators/PB103_is_playbook_has_unconnected_tasks.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    error_message = (
+        "Playbook conditional task with id:{task_id} has task with unreachable "
+        'next task condition "{next_task_branch}". Please remove this task or add '
+        "this condition to condition task with id:{task_id}."
+    )
+    related_field = "tasks"
+    is_auto_fixable = False
+
+    @staticmethod
+    def _is_unconnected_task(playbook: ContentTypes) -> bool:
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/validators/PB_validators/PB103_is_playbook_has_unconnected_tasks.py:27]:
```suggestion
    def is_unconnected_task(playbook: ContentTypes) -> bool:
```

File: demisto_sdk/commands/validate/validators/PB_validators/PB103_is_playbook_has_unconnected_tasks.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+    def is_valid(self, content_items: Iterable[ContentTypes]) -> List[ValidationResult]:
+        return [
+            ValidationResult(
+                validator=self,
+                message=self.error_message,
+                content_object=content_item,
+            )
+            for content_item in content_items
+            if (self._is_unconnected_task(content_item))
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/validators/PB_validators/PB103_is_playbook_has_unconnected_tasks.py:59]:
```suggestion
            if (self.is_unconnected_task(content_item))
```

Comment (YuvHayun) [inline demisto_sdk/commands/validate/validators/PB_validators/PB103_is_playbook_has_unconnected_tasks.py:59]: Any chance we're missing the orphan tasks here?
How do we format them into the error message?

File: demisto_sdk/commands/validate/tests/PB_validators_test.py [hunk 6]
```diff
@@ -215,3 +216,57 @@ def test_IsAskConditionHasUnhandledReplyOptionsValidator():
...
+            "message": {"replyOptions": ["yes"]},
+            "nexttasks": {"no": ["2"]},
+        },
+        "3": {
+            "id": "test task",
+            "type": "condition",
+            "message": {"replyOptions": ["yes"]},
+            "nexttasks": {"no": ["2"]},
+        },
+    }
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/tests/PB_validators_test.py:271]: you can also add it as a part of create_playbook_object.

File: demisto_sdk/commands/validate/tests/PB_validators_test.py [hunk 7]
```diff
@@ -215,3 +216,57 @@ def test_IsAskConditionHasUnhandledReplyOptionsValidator():
...
+    assert not IsPlaybookHasUnconnectedTasks().is_valid([playbook])
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/tests/PB_validators_test.py:272]: Let's also assert the message content.

**Review by YuvHayun — CHANGES_REQUESTED** [review]
Nice work!
See my comments.

File: demisto_sdk/commands/validate/validators/PB_validators/PB103_is_playbook_has_unconnected_tasks.py (hunk 5 above)
Comment (AradCarmi) [reply demisto_sdk/commands/validate/validators/PB_validators/PB103_is_playbook_has_unconnected_tasks.py:59]: WDYM? we are iterating over the playbook tasks and searching for the tasks that are not connected to the root.

File: demisto_sdk/commands/validate/tests/PB_validators_test.py [hunk 8]
```diff
@@ -222,3 +224,91 @@ def test_IsAskConditionHasUnhandledReplyOptionsValidator():
...
+            **{
+                "id": "test task",
+                "type": "condition",
+                "message": {"replyOptions": ["yes"]},
+                "nexttasks": {"no": ["2"]},
+                "task": {"id": "27b9c747-b883-4878-8b60-7f352098a63c"},
+                "taskid": "27b9c747-b883-4878-8b60-7f352098a632",
+            }
+        ),
+    }
```
Comment (YuvHayun) [inline demisto_sdk/commands/validate/tests/PB_validators_test.py:310]: put under create_playbook_object

**Review by YuvHayun — APPROVED** [review]

File: demisto_sdk/commands/validate/validators/PB_validators/PB103_is_playbook_has_unconnected_tasks.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+    @staticmethod
+    def is_unconnected_task(playbook: ContentTypes) -> set[Any]:
+        """Checks whether a playbook has an unconnected task.
+        Args:
+            - playbook (ContentTypes): The playbook to check.
+        Return:
+            - bool. True if the playbook has an unconnected task, False otherwise.
+        """
+        start_task_id = playbook.data.get("starttaskid")
```
[the remark this answers is not in the document]
Comment (AradCarmi) [reply demisto_sdk/commands/validate/validators/PB_validators/PB103_is_playbook_has_unconnected_tasks.py:31]: maybe on another PR.

File: demisto_sdk/commands/validate/tests/PB_validators_test.py [hunk 10]
```diff
@@ -222,3 +224,89 @@ def test_IsAskConditionHasUnhandledReplyOptionsValidator():
...
+                    "taskid": "27b9c747-b883-4878-8b60-7f352098a632",
+                },
+            },
+        ],
+    )
+    validation_results = IsPlaybookHasUnconnectedTasks().is_valid([playbook])
+    assert len(validation_results) == 0  # No validation results should be returned
+
+
+def test_is_playbook_has_unconnected_tasks_not_valid():
```
Comment (dantavori) [inline demisto_sdk/commands/validate/tests/PB_validators_test.py:263]:
```suggestion
def test_does_playbook_have_unconnected_tasks_not_valid():
```
etc

**Review by dantavori — APPROVED** [review]

**Pull request merged**